src/modules/linux-micro/kmod/kmod.c

```diff
@@ -0,0 +1,410 @@
+    else
+        fd = openat(dfd, file_name, O_RDONLY | O_CLOEXEC);
+
+    if (fd < 0) {
+        SOL_DBG("Could not open: %s", file_name);
+        return 0;
+    }
+
+    res = kmod_apply_file(kmod, fd);
+    close(fd);
+    return res;
+}
+
+static int
+kmod_conf_cmp(const void *data1, const void *data2)
+{
+    return strcmp((const char *)data1, (const char *)data2);
+}
+
+static int
+read_directory_to_vector(DIR *dir, struct sol_ptr_vector *vec)
+{
+    struct dirent ent, *entbuf = NULL;
+    char *itr;
+    uint16_t i;
+    int r;
+
+    r = readdir_r(dir, &ent, &entbuf);
+    while (r == 0 && entbuf) {
+        char *s, *dot = strrchr(ent.d_name, '.');
+
+        if (!dot || !streq(dot, ".conf")) {
+            SOL_DBG("File name does not end in '.conf', ignoring: %s",
+                ent.d_name);
+            goto next;
+        }
+
+        s = strdup(ent.d_name);
+        if (!s) {
+            SOL_WRN("Could not allocate memory to hold dir entry: %s",
+                ent.d_name);
+            goto str_err;
+        }
+
+        sol_ptr_vector_insert_sorted(vec, s, kmod_conf_cmp);
+
+next:
+        r = readdir_r(dir, &ent, &entbuf);
+    }
+
+    return 0;
+
+str_err:
+    SOL_PTR_VECTOR_FOREACH_IDX (vec, itr, i) {
+        free(itr);
+    }
+    sol_ptr_vector_clear(vec);
+    return -ENOMEM;
+}
+
+static void
+kmod_coldplug_find_devices(struct kmod_data *kdata, DIR *dir)
+{
+    struct dirent ent, *entbuf = NULL;
+    int r, dfd;
+
+    dfd = dirfd(dir);
+    if (dfd < 0) {
+        SOL_DBG("Could not obtain directory file descriptor");
+        return;
+    }
+
+    r = readdir_r(dir, &ent, &entbuf);
+    while (r == 0 && entbuf) {
+        struct stat st;
+
+        if (!strcmp(ent.d_name, ".") || !strcmp(ent.d_name, "..")) {
+            goto next;
+        }
+
+        if (fstatat(dfd, ent.d_name, &st, AT_SYMLINK_NOFOLLOW) == -1) {
+            SOL_ERR("Could not stat %s", ent.d_name);
+            goto next;
+        }
+
+        if (S_ISDIR(st.st_mode)) {
+            int nfd;
+            DIR *ndir;
+
+            nfd = openat(dfd, ent.d_name, O_RDONLY | O_CLOEXEC);
+            if (nfd < 0) {
+                SOL_ERR("Could not open sys directory %s - %s", ent.d_name,
+                    sol_util_strerrora(errno));
+                goto next;
+            }
+
+            ndir = fdopendir(nfd);
+            if (!ndir) {
+                SOL_ERR("Could not open sys directory %s", ent.d_name);
+                close(nfd);
+                goto next;
+            }
+
+            kmod_coldplug_find_devices(kdata, ndir);
+            closedir(ndir);
+        } else if (streq(ent.d_name, "modalias")) {
+            int fd;
+            char buff[PATH_MAX];
+
+            fd = openat(dfd, ent.d_name, O_RDONLY | O_CLOEXEC);
+            if (fd < 0) {
+                SOL_ERR("Could not obtain uevent file descriptor");
+                goto next;
+            }
+
+            if (read(fd, buff, sizeof(buff)) > 0) {
```

Comment:
Did you test this when building with `-O0`? Because you might be getting lucky here and `buff` might be 0-initialized, thus the string read from the file ends up being 0-terminated. To avoid problems, store the return value of the call to `read()`, and pass the slice with that as the length to `kmod_apply_value()`.